### ISSUE TITLE
Doc tests are not built with specified features

### ISSUE THREAD

**Issue by apasel422** [rust-lang/cargo#1277]
With https://github.com/apasel422/foo and `cargo 0.0.1-pre-nightly (2b9a3d6 2015-01-30 02:47:30 +0000)`

```sh
> cargo --test --features bar
   Compiling foo v0.0.1 (file:///home/andrew/foo)
     Running target/foo-d486511c73453e56

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured

   Doc-tests foo

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured
```

**Comment by apasel422** [comment]
The verbose output makes it clear that the `rustdoc` command is not being invoked with the correct features:
```sh
cargo test --features bar --verbose
   Compiling foo v0.0.1 (file:///home/andrew/foo)
     Running `rustc src/lib.rs --crate-name foo --crate-type lib -g --cfg feature="bar" -C metadata=7cf71ac9a7d9fc8e -C extra-filename=-7cf71ac9a7d9fc8e --out-dir /home/andrew/foo/target --emit=dep-info,link -L dependency=/home/andrew/foo/target -L dependency=/home/andrew/foo/target/deps`
     Running `rustc src/lib.rs --crate-name foo --crate-type lib -g --test --cfg feature="bar" -C metadata=d486511c73453e56 -C extra-filename=-d486511c73453e56 --out-dir /home/andrew/foo/target --emit=dep-info,link -L dependency=/home/andrew/foo/target -L dependency=/home/andrew/foo/target/deps`
     Running `/home/andrew/foo/target/foo-d486511c73453e56`

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured

   Doc-tests foo
     Running `rustdoc --test /home/andrew/foo/src/lib.rs --crate-name foo -L /home/andrew/foo/target -L /home/andrew/foo/target/deps --extern foo=/home/andrew/foo/target/libfoo-7cf71ac9a7d9fc8e.rlib`

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured
```

**Comment by apasel422** [comment]
https://github.com/rust-lang/cargo/blob/master/src/cargo/ops/cargo_test.rs#L62-L83 doesn't contain any feature-handling code. Compare with https://github.com/rust-lang/cargo/blob/master/src/cargo/ops/cargo_rustc/mod.rs#L514-L521.

**Comment by apasel422** [comment]
I'm trying to fix this myself, but I don't see how to get access to the list of features without a `Context`.

**Comment by alexcrichton** [comment]
You'll probably have to add a field to the `Compilation` and just fill it in as part of one of the steps in the `cargo_rustc` module (the `Context` can't live past that module).

**Comment by apasel422** [comment]
I'm starting to think there might also be a bug in rustdoc:
```sh
> rustdoc --cfg feature=bar baz
fgspec:1:9: 1:12 error: unexpected token: `bar`
cfgspec:1 feature=bar
                  ^~~
thread '<unnamed>' panicked at 'Box<Any>', /home/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libsyntax/diagnostic.rs:92
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: "rustc failed"', /home/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libcore/result.rs:743
thread '<main>' panicked at 'called `Option::unwrap()` on a `None` value', /home/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libcore/option.rs:362
```

**Comment by apasel422** [comment]
This depends on rust-lang/rust#22131.

**Comment by alexcrichton** [comment]
I belive this was fixed back in #1284
